app/controllers/application_controller/filter.rb

```diff
@@ -947,7 +947,11 @@ def adv_search_build(model)
     # Restore @edit hash if it's saved in @settings
     @expkey = :expression                                               # Reset to use default expression key
     if session[:adv_search] && session[:adv_search][model.to_s]
-      @edit = copy_hash(session[:edit])
```

Comment:
The very reason the crash happened was because `@edit[@expkey]` was `nil`.
Check out line#977, which is expecting `@edit[@expkey]` to have a valid value - 
```ruby
@edit[@expkey][:exp_table] = exp_build_table(@edit[@expkey][:expression]) 
```

So the conditional that I have added -
```ruby
if session[:adv_search][model.to_s][@expkey]
```
specifically looks for that case before it decides to do a `copy_hash` on `session[:adv_search][model.to_s]`, which I think is reasonable.

It's hard to predict beforehand how many types of use cases we expect to see here (maybe you have an idea), but for now I think we have covered what the BZ demands and have not broken what was fixed in #776.

Thanks for exploring further.